test-bot: check all dependents for broken dylibs

apjanke opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031). **Answer:** *No. This is mostly in test-bot so we don't have a harness to test it. And `brew linkage` depends on having a formula installed, so we can't test that reproducibly without interacting with the Homebrew installation.*
- [X] Have you successfully ran `brew tests` with your changes locally?

Goes with Homebrew/homebrew-dev-tools/pull/5.
Addresses #60 

##  Description  ##

Has `test-bot` check all dependents of changed formulae for dylib link breakage, independent of whether they have a `test do` block.

Pulls 'brew linkage' in to main brew repo as a dev-cmd, and has test-bot use it to detect dylib breakage, which usually means a revision bump is needed. Checks all dependents, not just those with a 'test do' block defined, since we can do this without formula support.

This follows up #60 ("Preventing formulae updates from breaking their dependencies") and partially addresses its concerns. Doesn't fully fix them, because the decoupling between repos means that test-bot still won't be able to validate cross-repo library dependencies. It will take manual work on the PR author's part to make sure separate PRs are made for revision bumps dependent repos and that all formulae are included.

##   Testing  ##

You can see an example of the new output at https://gist.github.com/apjanke/77f7e593c922a58408bdc1f92964bcec.

This was produced with `brew test-bot --keep-logs --junit --local --verbose --skip-homebrew --skip-setup icu4c  | tee test-bot-with-linkage.txt` run against the current repo states.

You can see it catching dylib breakage that's still in some of `icu4c`'s dependents. Some of these were not caught by the existing tests, because they don't define a `test do` block, so they're not "testable". This would show up in Jenkins as a failed test step.

```

==> brew linkage --test gptfdisk
No broken dylib links
==> brew linkage --test libfolia
Broken dylib links:
  /usr/local/opt/icu4c/lib/libicudata.56.1.dylib
  /usr/local/opt/icu4c/lib/libicui18n.56.dylib
  /usr/local/opt/icu4c/lib/libicuio.56.dylib
  /usr/local/opt/icu4c/lib/libicuuc.56.dylib
==> FAILED
==> brew linkage --test ucto
Broken dylib links:
  /usr/local/opt/icu4c/lib/libicudata.56.1.dylib
  /usr/local/opt/icu4c/lib/libicui18n.56.dylib
  /usr/local/opt/icu4c/lib/libicuio.56.dylib
  /usr/local/opt/icu4c/lib/libicuuc.56.dylib
==> FAILED
==> brew linkage --test vislcg3
Broken dylib links:
  /usr/local/opt/icu4c/lib/libicui18n.56.dylib
  /usr/local/opt/icu4c/lib/libicuio.56.dylib
  /usr/local/opt/icu4c/lib/libicuuc.56.dylib
  @rpath/libcg3.0.dylib
```

##  Changes and Compatibility  ##

With this, `brew linkage` is now a dev-cmd inside main `brew`, instead of a regular cmd in `homebrew/dev-tools`. Users will need to have `HOMEBREW_DEVELOPER` enabled to make it visible. Users who had `homebrew/dev-tools` tapped but didn't have `HOMEBREW_DEVELOPER` set will lose it.

I chose to put it in `dev-cmd` because the move was done for `test-bot`'s use, and it was originally a "dev" tool. I think it'd be fine to just have in `cmd`, too, to avoid breaking things for anybody, if other maintainers think that would be more appropriate.

Any other taps which define a `linkage` command will now have it masked when `HOMEBREW_DEVELOPER` is enabled. (Or always, if we choose to make it a regular `cmd`.)

UniqMartin commented:
> ```
> ==> brew linkage --test vislcg3
> Broken dylib links:
>   /usr/local/opt/icu4c/lib/libicui18n.56.dylib
>   /usr/local/opt/icu4c/lib/libicuio.56.dylib
>   /usr/local/opt/icu4c/lib/libicuuc.56.dylib
>   @rpath/libcg3.0.dylib
>  ```

To avoid getting false positives like in the quoted example, we either need to teach `brew linkage` how to properly resolve the special linker path prefixes `@executable_path/`, `@loader_path/`, and `@rpath/` (resolving all of them correctly is non-trivial and sometimes even impossible) or we need to ignore those entries as far as our check for brokenness is concerned (we're building most things with absolute paths anyway, so we're not missing much). I'm strongly suggesting the latter option.

Otherwise generally :+1: on this.

UniqMartin commented on Library/Homebrew/cmd/test-bot.rb:
```diff
@@ -583,7 +583,7 @@ def formula(formula_name)
         end
         test "brew", "test", "--verbose", formula_name if formula.test_defined?
-        testable_dependents.each do |dependent|
+        dependents.each do |dependent|
```
Unlike the name suggests, `testable_dependents` are not merely dependents with a test, but dependents that have a test and are bottled (see a bit further up). Switching to `dependents` potentially means that unbottled dependents will be built from source, just for the sake of checking their linkage, which is likely to be alright due to the from-source build.

apjanke replied:
Good point. Yeah, testing linkage on non-bottled formulae is probably pointless. I'll modify it to do only bottled ones.

mikemcquaid replied:
It'd be useful but it's too slow.

apjanke replied:
Switched to doing just the bottled dependents.

mikemcquaid commented on Library/Homebrew/dev-cmd/linkage.rb:
```diff
@@ -0,0 +1,118 @@
+    end
+  end
+
+end
```
newline

apjanke replied:
Fixed.

mikemcquaid commented:
Seems good, thanks for this 👍

apjanke commented:
Switched to testing bottled dependents only.

I modified it to treat install paths starting with `@` variables as a different category, and to not consider them as broken links. New output:

```
$ brew linkage vislcg3
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libc++.1.dylib
Variable-referenced libraries:
  @rpath/libcg3.0.dylib
Missing libraries:
  /usr/local/opt/icu4c/lib/libicui18n.56.dylib
  /usr/local/opt/icu4c/lib/libicuio.56.dylib
  /usr/local/opt/icu4c/lib/libicuuc.56.dylib
$ brew linkage --test vislcg3
Broken dylib links:
  /usr/local/opt/icu4c/lib/libicui18n.56.dylib
  /usr/local/opt/icu4c/lib/libicuio.56.dylib
  /usr/local/opt/icu4c/lib/libicuuc.56.dylib
[✘ /usr/local on ⇄ detect-broken-dylibs ±]
```

I read a couple articles on the `@xxx/...` install path form, including [Mike Ash's Q&A](https://www.mikeash.com/pyblog/friday-qa-2009-11-06-linking-and-install-names.html), `man dyld`, and `man install_name_tool`. I think validating these is impossible for us, since they are relative paths that are evaluated, at least in part, relative to the program or framework that is linking to the library, which we don't know when we're examining the library on its own. And per the docs, an `@rpath` that does not resolve is not an error.

(To be strict, maybe we should be checking for unresolved symbols before complaining about referenced library paths that don't resolve. But I think 99% of the time they'll be an error for us regardless, so we probably don't have to check the symbols, too.)

UniqMartin commented on Library/Homebrew/dev-cmd/linkage.rb:
```diff
@@ -0,0 +1,127 @@
+      @keg.find do |file|
+        next unless file.dylib? || file.mach_o_executable? || file.mach_o_bundle?
+        file.dynamically_linked_libraries.each do |dylib|
+          if !dylib.empty? && dylib[0] == "@"
```
I find `dylib.start_with?("@")` easier to parse.

apjanke replied:
Agreed.

UniqMartin commented on Library/Homebrew/dev-cmd/linkage.rb:
```diff
@@ -0,0 +1,127 @@
+
+    def display_test_output
+      if @broken_dylibs.empty?
+        puts "No broken dylib links"
```
This information is already encoded in the exit status and it seems to be customary for most CLI programs to output nothing in this case. I think it is sufficient to output only when there is something to report on.

apjanke replied:
Generally yes. But most other `brew` subcommands, including diagnostic ones like `brew test`, `brew style`, and `brew doctor`, include some output even in the normal "no problems" case. Did this to be consistent.

UniqMartin commented on Library/Homebrew/dev-cmd/linkage.rb:
```diff
@@ -0,0 +1,127 @@
+      if @broken_dylibs.empty?
+        puts "No broken dylib links"
+      else
+        puts "Broken dylib links:"
```
I find it confusing that the terminology is different depending on whether `--test` was supplied or not. Here it's “Broken” but otherwise it's “Missing”. Could also be worthwhile to split `LinkageChecker#display_normal_output` into smaller methods and reuse the code here.

apjanke replied:
Different terminology for the different use contexts. I didn't want to change the wording of the existing `brew linkage` behavior. But when using it in `test-bot`, it's more about detecting broken links which need to be changed than a "missing" library which should be installed. I don't think it would be a big deal to change it to "Missing libraries" for the `--test` version, too.

apjanke replied:
Yeah, if you're confused about it, that's probably a bad sign. And it is outputting exactly the same info in both cases; probably makes more sense to use the exact same label. I've changed it to do "Missing libraries" in both normal and `--test` version.

UniqMartin replied:
Though now that I know your thought process, it kinda makes sense to me. Thanks for taking the time to explain and changing it nonetheless. :bow:

xu-cheng commented on Library/Homebrew/dev-cmd/linkage.rb:
```diff
@@ -0,0 +1,127 @@
+    end
+  end
+
+  class LinkageChecker
```
It may be worth to add this checker to `formula_cellar_checks.rb`, i.e. checking linkage of the formula itself rather than the dependents.

apjanke replied:
Probably, similar to how we do a check like `brew missing` there. Mind if I put it off to a later PR, though?

xu-cheng replied:
No problem. But you may want to put this class in a standalone file if it will be used in other part of core in the future.

apjanke commented:
@mikemcquaid You're welcome!

apjanke commented:
And here's what the results look like, from https://github.com/Homebrew/homebrew-core/pull/427.

<img width="785" alt="screen shot 2016-04-18 at 2 14 32 pm" src="https://cloud.githubusercontent.com/assets/2618447/14614639/0adac93a-0570-11e6-8155-14a6edef94ed.png">

<img width="446" alt="screen shot 2016-04-18 at 2 14 47 pm" src="https://cloud.githubusercontent.com/assets/2618447/14614643/0e418b4a-0570-11e6-8ff6-a1b7772e8707.png">

Gonna go make a revision-bump PR now... :)

apjanke commented:
It's working! We've caught dylib breakage in PRs a couple times this last week. Example: https://github.com/Homebrew/homebrew-core/pull/621

ahundt commented:
cool! so is this functionality integrated or not merged? confused by the red closing of the pull request.

apjanke commented:
Thanks! It's integrated and running now (at least for in-tap dependencies), and we've already caught and fixed a couple broken bottles with it.

Most of our closed pull requests for things that get accepted are still red "closed" instead of purple "merged" because we usually use `brew pull` instead of GitHub's "merge" button to do the merging. This allows us to edit commit messages, link commit messages back to PRs, include test-bot-generated bottle pulls, squash commits, do local testing, and so on.

If the message says something with a commit identifier like "apjanke closed this in da34fba 14 days ago", that means it was accepted and merged. If the message just says "apjanke closed this 14 days ago" that means it was probably rejected or superseded, and there should be an explicit comment in the discussion indicating why it was rejected.

Sorry if this is confusing.

apjanke commented:
Maybe GitHub could make a special "Mark as manually merged" button. But maybe that would break their model with tight coupling between commit history and issues.

ahundt commented:
ah ok I see now, thanks for the clarification! It would be nice if github allowed things to be made clearer more easily...
